### PR TITLE
Speed-up casts to FP8

### PR DESCRIPTION
**Pull request by malfet**
Unlike half/bfloat16 casts, where entire model is cast to half-precision floats, only parts of the network can be in float8 and therefore performance of the casts is important.

Speedup casts by implementing non-dynamically castable variants using new refactored `gpu_kernel_nocast` template.

Mesaure performance using the following script:
```python
import torch

def run_cast_bench(size=(10000, 10000), src_dtype=torch.float16, dtype=torch.float8_e5m2):
    x=torch.rand(size, device="cuda", requires_grad=False, dtype=src_dtype)
    z=torch.empty(size, device="cuda", dtype=dtype, requires_grad=False)
    with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CUDA]) as prof:
        z.copy_(x)
    rc=prof.key_averages()
    print(f"Running bench for src_dtype={src_dtype} dst_dtype={dtype} cuda_time={rc[1].cuda_time}")

if __name__ == "__main__":
    for dtype in [torch.float8_e5m2, torch.float8_e4m3fn]:
        run_cast_bench(src_dtype=torch.half, dtype=dtype)
        run_cast_bench(src_dtype=torch.float, dtype=dtype)
        run_cast_bench(src_dtype=torch.bfloat16, dtype=dtype)
```

Below are before and after results:
|  Cast type | After | Before |
| ---------- | ------ | ----- |
| fp32->e5m2 | 228 us | 336 us|
| fp16->e5m2 | 150 us | 323 us|
| bf16->e5m2 | 150 us | 322 us|
| fp32->e4m3 | 227 us | 331 us|
| fp16->e4m3 | 148 us | 318 us|
| bf16->e4m3 | 149 us | 318 us|

Skip the optimizations on ROCm platform
TODO:
 - Investigate why `__nv_cvt` intrinsics defined in https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__FP8__MISC.html end up being slower

### CODE REVIEW

File: aten/src/ATen/native/cuda/Copy.cu [hunk 1]
```diff
@@ -19,20 +19,82 @@
...
 namespace at::native {
 
 void neg_kernel_cuda(TensorIteratorBase &iter);
 void conj_kernel_cuda(TensorIteratorBase &iter);
 
+void float8_copy_kernel_cuda(TensorIteratorBase &iter) {
+  ScalarType dtype = iter.dtype(0);
+  ScalarType other_dtype = iter.dtype(1);
+  if (dtype == kFloat8_e4m3fn) {
+    if (other_dtype == kFloat) {
```
Comment (drisspg) [inline aten/src/ATen/native/cuda/Copy.cu:36]: nit: would switch make this easier to read?

File: aten/src/ATen/native/cuda/Copy.cu [hunk 2]
```diff
@@ -19,20 +19,82 @@
...
+#if 0
+           const auto x =  __nv_cvt_halfraw_to_fp8(static_cast<__half>(value), __NV_NOSAT, __NV_E5M2);
+           return Float8_e5m2(x, Float8_e5m2::from_bits());
+#else
+           return Float8_e5m2(value);
+#endif
+       });
+    } else if (dtype == kFloat8_e5m2 && other_dtype == kBFloat16) {
+       gpu_kernel_nocast(iter, [] GPU_LAMBDA(BFloat16 value) {
+#if 0
```
Comment (drisspg) [inline aten/src/ATen/native/cuda/Copy.cu:72]: Nit: should we create the macro now and just set it to 0

File: aten/src/ATen/native/cuda/Copy.cu [hunk 3]
```diff
@@ -19,20 +19,82 @@
...
+    if (other_dtype == kFloat) {
+       gpu_kernel_nocast(iter, [] GPU_LAMBDA(float value) {
+#if 0
+           const auto x =  __nv_cvt_float_to_fp8(value, __NV_NOSAT, __NV_E5M2);
+           return Float8_e5m2(x, Float8_e5m2::from_bits());
+#else
+           return Float8_e5m2(value);
+#endif
+       });
+    } else if (other_dtype == kHalf) {
```
Comment (drisspg) [inline aten/src/ATen/native/cuda/Copy.cu:61]: are fp64 casts supported?

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 4]
```diff
@@ -310,4 +333,4 @@ void gpu_kernel_multiple_outputs(TensorIteratorBase& iter, const func_t& f) {
   gpu_kernel_multiple_outputs_impl(iter, f);
 }
 
-}} //namespace at::native
+} //namespace at::native
```
Comment (drisspg) [inline aten/src/ATen/native/cuda/Loops.cuh:336]: was this intentional?

**Review by drisspg — APPROVED** [review]
I think this looks good and the performance wins look great.

I see that 
``` Python
import torch
from transformer_nuggets.utils.tracing import LoggingMode

a = torch.rand(2, dtype=torch.float16, device='cuda')
with LoggingMode():
    a.to(torch.float8_e4m3fn)
    
```

prints
`$1: f8e4[2] = aten._to_copy.default($0, dtype=torch.float8_e4m3fn)`

would we hit this path using the `.to` function?

File: aten/src/ATen/native/cuda/Copy.cu (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/cuda/Copy.cu:72]: Sorry, those supposed to be guarded by the macro, let me replace this one..

File: aten/src/ATen/native/cuda/Copy.cu (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/native/cuda/Copy.cu:36]: Can you please elaborate what do you mean? You have top-level ifs for casts from float, bfloat16 and float16?

**Comment by drisspg** [comment]
Ran `python benchmarks/bench_linear_float8.py --use_ts -o benchmarks/data/float8_sweep_cast_update` before and after this pr:
| name      | shape               | ref_dtype      |   pt_fp8_time_sec |   pt_fp8_time_sec_updated |   % Improvement |
|:----------|:--------------------|:---------------|------------------:|--------------------------:|----------------:|
| attn.wqkv | (16384, 8192, 1280) | torch.bfloat16 |        0.00416412 |                0.00390124 |         6.31303 |
| attn.w0   | (16384, 1024, 8192) | torch.bfloat16 |        0.00361837 |                0.00335387 |         7.30976 |
| ffn.w13   | (16384, 8192, 7168) | torch.bfloat16 |        0.0107657  |                0.0102925  |         4.39569 |
| ffn.w2    | (16384, 3584, 8192) | torch.bfloat16 |        0.00660132 |                0.00621062 |         5.91851 |
| attn.wqkv | (16384, 8192, 1280) | torch.float16  |        0.00423888 |                0.00398655 |         5.95275 |
| attn.w0   | (16384, 1024, 8192) | torch.float16  |        0.00367196 |                0.0034114  |         7.09605 |
| ffn.w13   | (16384, 8192, 7168) | torch.float16  |        0.0107916  |                0.010287   |         4.67511 |
| ffn.w2    | (16384, 3584, 8192) | torch.float16  |        0.00662924 |                0.00625864 |         5.59038 |


Looks like this is producing very meaningful speedups!

File: aten/src/ATen/native/cuda/Loops.cuh (hunk 4 above)
Comment (malfet) [reply aten/src/ATen/native/cuda/Loops.cuh:336]: No, I like nested namespaces, it's a neat C++17 feature :)

**Pull request closed without merging**